rust/parser/typeql.pest

```diff
@@ -0,0 +1,327 @@
+variable_thing_any = { variable_thing | variable_relation | variable_attribute }
+
+variable_thing = { VAR_ ~ ISA_ ~ type_ ~ ( "," ~ attributes )?
+                 | VAR_ ~ IID ~ IID_ ~ ( "," ~ attributes )?
+                 | VAR_ ~ attributes
+                 }
+
+variable_relation = { VAR_? ~ relation ~ ISA_ ~ type_ ~ ( "," ~ attributes )?
+                    | VAR_? ~ relation ~ attributes? }
+
+variable_attribute = { VAR_? ~ predicate ~ ISA_ ~ type_ ~ ( "," ~ attributes )?
+                     | VAR_? ~ predicate ~ attributes? }
+
+// RELATION CONSTRUCT ==========================================================
+
+relation = { "(" ~ role_player ~ ( "," ~ role_player )* ~ ")" } // A list of role players in a Relations
+role_player = { type_ ~ ":" ~ player | player } // The Role type and player variable or just the player variable
+player = { VAR_ } // A player is just a variable
+
+// ATTRIBUTE CONSTRUCT =========================================================
+
+attributes = { attribute ~ ( "," ~ attribute )* }
+attribute = { HAS ~ label ~ ( VAR_ | predicate ) | HAS ~ VAR_ } // ownership by labeled variable or value or just value
+
+// ATTRIBUTE VALUATION CONSTRUCTS ==============================================
+
+predicate = { value | predicate_equality ~ predicate_value | predicate_substring ~ STRING_ }
+
+predicate_equality = { EQ | NEQ | GTE | GT | LTE | LT }
+predicate_substring = { CONTAINS | LIKE }
+
+predicate_value = { value | VAR_ }
+
+// SCHEMA CONSTRUCT =============================================================
+
+schema_rule_declaration = { RULE ~ label }
+schema_rule = { RULE ~ label ~ ":" ~ WHEN ~ "{" ~ patterns ~ "}" ~ THEN ~ "{" ~ variable_thing_any ~ ";" ~ "}" }
+
+// TYPE, LABEL AND IDENTIFIER CONSTRUCTS =======================================
+
+type_any = { type_scoped | type_ | VAR_ }
+type_scoped = { label_scoped | VAR_ }
+type_ = { label | VAR_ } // A type_can be a label or variable
+
+label_any = { label_scoped | label }
+label_scoped = { LABEL_SCOPED_ }
+label = { LABEL_ | schema_native | type_native | unreserved }
+
+// LITERAL INPUT VALUES =======================================================
+
+schema_native = { RULE }
+
+type_native = { THING | ENTITY | ATTRIBUTE | RELATION | ROLE }
+
+value_type = { LONG | DOUBLE | STRING | BOOLEAN | DATETIME }
+value = { STRING_ | DATETIME_ | DATE_ | DOUBLE_ | LONG_ | BOOLEAN_ }
+regex = { STRING_ }
+
+// UNRESERVED KEYWORDS =========================================================
+// Most of TypeQL syntax should not be reserved from being used as identifiers
+
+unreserved = { VALUE | MIN | MAX| MEDIAN | MEAN | STD | SUM | COUNT | GET | SORT | LIMIT | OFFSET | GROUP | CONTAINS }
+
+
+// TYPEQL SYNTAX KEYWORDS =======================================================
+
+// QUERY COMMAND KEYWORDS
+
+MATCH = @{ "match" ~ WB }
+GET = @{ "get" ~ WB }
+DEFINE = @{ "define" ~ WB }
+UNDEFINE = @{ "undefine" ~ WB }
+INSERT = @{ "insert" ~ WB }
+DELETE = @{ "delete" ~ WB }
```

Comment:
`@{ ... }` denotes an atomic rule. An atomic rule contains no implicit whitespace, and cannot have children.
`WB` is a word boundary, which performs a lookahead expecting whitespace, comments, or punctuation. Without checking wor word boundaries, since whitespace in `~` is optional, `matchmovie sub$t;` would be a valid query, equivalent to `match movie sub $t;`.